mlflow/sklearn/__init__.py

```diff
@@ -455,6 +461,265 @@ def load_model(model_uri):
+        * storing the dataset row samples, we need log it into metric_info artifact file.
+    (5) metric_api_call_arg_dict_list_map, it is a map, key is metric api function name,
+       and value is a list of call arguments dict.
+    """
+
+    fit_scope_key = 'fit'
+    eval_and_log_metrics_scope_key = 'eval_and_log_metrics'
+    model_score_scope_key = 'model_score'
+
+    def __init__(self):
+        self.model_id_to_run_id_map = {}
+        self.pred_result_id_to_dataset_name_and_model_id = {}
+        self.scope_flag_dict = {
+            self.fit_scope_key: False,
+            self.eval_and_log_metrics_scope_key: False,
+            self.model_score_scope_key: False
+        }
+        self.eval_dataset_info_map = defaultdict(lambda: defaultdict(list))
+        self.metric_api_call_arg_dict_list_map = defaultdict(list)
+
+    def call_scope(self, scope_key):
+        class CallScope:
+            def __enter__(inner_self):
+                self.scope_flag_dict[scope_key] = True
+
+            def __exit__(inner_self, exc_type, exc_val, exc_tb):
+                self.scope_flag_dict[scope_key] = False
+
+        return CallScope()
+
+    def get_run_id_for_model(self, model):
+        return self.model_id_to_run_id_map.get(model._mlflow_uuid, None)
+
+    def should_log_eval_metrics(self, metric_value=None):
+        """
+        Test whether we should log metric. will check metric value types (if provided)
+        and scope flags
+        """
+        return (np.isscalar(metric_value) if metric_value is not None else True) and \
+            not self.scope_flag_dict[self.fit_scope_key] and \
+            not self.scope_flag_dict[self.eval_and_log_metrics_scope_key] and \
+            not self.scope_flag_dict[self.model_score_scope_key]
+
+    def register_model(self, model, run_id):
+        """
+        In `patched_fit`, we need register the model with the run_id used in `patched_fit`
+        So that in following metric autologging, the metric will be logged into the registered
+        run_id
+        """
+        model._mlflow_uuid = uuid4()
+        self.model_id_to_run_id_map[model._mlflow_uuid] = run_id
+
+    @staticmethod
+    def gen_name_with_index(name, index):
+        assert index >= 0
+        if index == 0:
+            return name
+        else:
+            return f'name_{index + 1}'
+
+    @staticmethod
+    def gen_dataset_sample_rows(eval_dataset):
+        n_sample = max(len(eval_dataset), INPUT_EXAMPLE_SAMPLE_ROWS)
+        return eval_dataset[:n_sample]
+
+    def register_eval_dataset(self, model, eval_dataset):
+        """
+        Register eval dataset into eval_dataset_info_map, it will do:
+         1. inspect eval dataset var name.
+         2. check whether eval_dataset_info_map already registered this eval dataset.
+            will check by object id and comparing row samples.
+            Note: only check object id is not sufficient, id may be reused.
```

Comment:
Now that we use weakref.finalize, do we still need to do the sampling?

Reply:
(1) The finalizer callback is applied on "prediction_result_object_id", but here it refer to the id of "eval_dataset_object". The two kinds of objects are different.

(2) for "eval_dataset_object", we can also register finalizer callback, but the callback cleanup logic will be complex (we need cleanup/rename related ids scattered in `eval_dataset_info_map`)

(3) The eval_dataset row samples is useful, prepared to be logged in artifacts (as discussed before), and here the row samples can be used to double check (avoid wrong mapping caused by ID reusing, and the row samples checking here is short-circuit so it won't affect performance)

Reply:
Sounds good - let's keep the sampling!

Reply:
Actually, do we want to log eval datasets as artifacts? They may contain sensitive information. We have a config flag for `log_input_examples` in autologging based on the training data. Do we need to log additional eval artifacts each time an eval dataset is passed to an sklearn metric? Seems like that might be overkill.

Reply:
ok. What about also checking `log_input_examples` param, if on then we also log sample rows for eval_dataset, otherwise do not log.

Reply:
seems like for most cases, for each model, user will score on one evaluation dataset. We can initially skip log additional eval row samples artifacts, if user requires this, we can add it in future version.

Reply:
+1 for not logging row samples for each metric. If anything, it would be more helpful to log a map from metric names to full API calls.

E.g. `accuracy_score_eval_X -> accuracy_score(eval_X, eval_y, normalize=True, ...)`

Reply:
I suggest prepend a prefix like `metric_call_{index}' before all autologged metric, like:
![image](https://user-images.githubusercontent.com/19235986/126188020-94b06c55-6f9b-40f2-b3d7-0c51806c3a95.png)
This will help sorting when displayed in the UI.

Reply:
@harupy 
The semantic now (proposed by @dbczumar) is: we create an increasing index when we detect a metric call from user code, and log the metric, and in artifact, also log the metric call command like: `logged_metric_key -> accuracy_score(eval_X, eval_y, normalize=True, ...)`

So if it is such semantic, I still prefer the key format of "{metric_name}_on_{xx_dataset}"

Reply:
Sounds good!

Reply:
In the runs table, how do logged metrics look like?

Reply:
This is how the `<call identifier>_<metric name>_<dataset name>` format looks like in the runs table:
![image](https://user-images.githubusercontent.com/17039389/126191498-2c1ffedf-5156-4d2f-bedf-cd2ed6293478.png)

Reply:
I agree with @harupy. I don't think we should add another `metric_call` prefix.